File: backend/apps/bank-service/src/services/bank.service.ts
/**
 * Bank Service - Business Logic for Bank Module
 */
import { BankCaseStatus, PrismaClient } from '@prisma/client';
import { createServiceLogger } from '../../../../libs/logging/logger.service';

const prisma = new PrismaClient();
const logger = createServiceLogger('bank-service');

export class BankService {
    /**
     * Get all banks
     */
    async getAllBanks(includeInactive = false) {
        return prisma.bank.findMany({
            where: includeInactive ? {} : { status: 'ACTIVE' },
            include: {
                _count: {
                    select: { branches: true }
                }
            },
            orderBy: { bankName: 'asc' }
        });
    }

    /**
     * Get bank by ID with branches and services
     */
    async getBankById(id: string) {
        return prisma.bank.findUnique({
            where: { id },
            include: {
                branches: true,
                services: true
            }
        });
    }

    /**
     * Register a new bank
     */
    async createBank(data: {
        bankName: string;
        bankCode: string;
        headOfficeLocation: string;
        contactEmail?: string;
        contactPhone?: string;
    }) {
        return prisma.bank.create({
            data: {
                ...data,
                status: 'ACTIVE'
            }
        });
    }

    /**
     * Update bank info
     */
    async updateBank(id: string, data: any) {
        return prisma.bank.update({
            where: { id },
            data
        });
    }

    /**
     * Branch Management
     */
    async createBranch(bankId: string, data: {
        branchName: string;
        district: string;
        sector: string;
        address: string;
        contactPhone?: string;
    }) {
        return prisma.bankBranch.create({
            data: {
                ...data,
                bankId,
                status: 'ACTIVE'
            }
        });
    }

    async getBranchesByBank(bankId: string) {
        return prisma.bankBranch.findMany({
            where: { bankId },
            orderBy: { branchName: 'asc' }
        });
    }

    /**
     * Bank Services
     */
    async addService(bankId: string, data: {
        serviceName: string;
        description?: string;
    }) {
        return prisma.bankService.create({
            data: {
                ...data,
                bankId,
                enabled: true
            }
        });
    }

    async toggleService(serviceId: string, enabled: boolean) {
        return prisma.bankService.update({
            where: { id: serviceId },
            data: { enabled }
        });
    }

    /**
     * Bank Cases (Complaints)
     */
    async submitBankCase(data: {
        bankId: string;
        branchId: string;
        serviceId: string;
        submitterId: string;
        description: string;
        evidenceUrl?: string;
    }) {
        // Generate a unique reference: BANK-YYYYMMDD-XXXX
        const dateStr = new Date().toISOString().slice(0, 10).replace(/-/g, '');
        const randomStr = Math.random().toString(36).substring(2, 6).toUpperCase();
        const reference = `BANK-${dateStr}-${randomStr}`;

        return prisma.bankCase.create({
            data: {
                ...data,
                caseReference: reference,
                status: 'RECEIVED'
            }
        });
    }

    async getCasesBySubmitter(submitterId: string) {
        return prisma.bankCase.findMany({
            where: { submitterId },
            include: {
                bank: true,
                branch: true,
                service: true
            },
            orderBy: { createdAt: 'desc' }
        });
    }

    async getCasesByBranch(branchId: string) {
        return prisma.bankCase.findMany({
            where: { branchId },
            include: {
                submitter: {
                    select: { name: true, phone: true }
                },
                service: true
            },
            orderBy: { createdAt: 'desc' }
        });
    }

    async updateCaseStatus(caseId: string, status: BankCaseStatus, performedBy: string, notes?: string) {
        return prisma.$transaction(async (tx) => {
            const updatedCase = await tx.bankCase.update({
                where: { id: caseId },
                data: { status }
            });

            await tx.bankCaseUpdate.create({
                data: {
                    caseId,
                    performedBy,
                    action: status,
                    notes
                }
            });

            return updatedCase;
        });
    }

    async getCaseDetails(caseId: string) {
        return prisma.bankCase.findUnique({
            where: { id: caseId },
            include: {
                bank: true,
                branch: true,
                service: true,
                submitter: {
                    select: { name: true, phone: true, email: true }
                },
                updates: {
                    include: {
                        performer: {
                            select: { name: true, role: true }
                        }
                    },
                    orderBy: { createdAt: 'asc' }
                }
            }
        });
    }
}

export const bankService = new BankService();

File: frontend/lib/bank/services/bank_service.dart
import '../../shared/services/api_client.dart';
import '../models/bank_models.dart';

class BankService {
  static final BankService _instance = BankService._();
  BankService._();
  static BankService get instance => _instance;

  /**
   * Admin Management API
   */

  // Get all banks
  Future<ApiResponse<List<BankModel>>> getAllBanks() async {
    final response = await apiClient.get<dynamic>('/banks');
    if (response.isSuccess && response.data != null) {
      final List<dynamic> data = response.data is List ? response.data : response.data['data'] ?? [];
      final banks = data.map((json) => BankModel.fromJson(json)).toList();
      return ApiResponse.success(banks);
    }
    return ApiResponse.error(response.error ?? 'Failed to load banks');
  }

  // Register Bank
  Future<ApiResponse<BankModel>> registerBank(Map<String, dynamic> data) async {
    final response = await apiClient.post('/banks/register', data);
    if (response.isSuccess && response.data != null) {
      return ApiResponse.success(BankModel.fromJson(response.data));
    }
    return ApiResponse.error(response.error ?? 'Failed to register bank');
  }

  // Get details (Branches/Services)
  Future<ApiResponse<Map<String, dynamic>>> getBankDetails(String bankId) async {
    final response = await apiClient.get('/banks/$bankId');
    if (response.isSuccess && response.data != null) {
      return ApiResponse.success(response.data);
    }
    return ApiResponse.error(response.error ?? 'Failed to load bank details');
  }

  // Add Branch
  Future<ApiResponse<BranchModel>> addBranch(String bankId, Map<String, dynamic> data) async {
    final response = await apiClient.post('/banks/$bankId/branches', data);
    if (response.isSuccess && response.data != null) {
      return ApiResponse.success(BranchModel.fromJson(response.data));
    }
    return ApiResponse.error(response.error ?? 'Failed to add branch');
  }

  // Add Service
  Future<ApiResponse<BankServiceModel>> addService(String bankId, Map<String, dynamic> data) async {
    final response = await apiClient.post('/banks/$bankId/services', data);
    if (response.isSuccess && response.data != null) {
      return ApiResponse.success(BankServiceModel.fromJson(response.data));
    }
    return ApiResponse.error(response.error ?? 'Failed to add service');
  }

  // Toggle Service
  Future<ApiResponse<bool>> toggleService(String serviceId, bool enabled) async {
    final response = await apiClient.patch('/banks/services/$serviceId', {'enabled': enabled});
    if (response.isSuccess) {
      return ApiResponse.success(true);
    }
    return ApiResponse.error(response.error ?? 'Update failed');
  }

  /**
   * Complaint API (Citizen Side)
   */

  // Submit Complaint
  Future<ApiResponse<BankCaseModel>> submitComplaint(Map<String, dynamic> data) async {
    final response = await apiClient.post('/banks/cases', data);
    if (response.isSuccess && response.data != null) {
      return ApiResponse.success(BankCaseModel.fromJson(response.data));
    }
    return ApiResponse.error(response.error ?? 'Complaint submission failed');
  }

  // Get My Complaints
  Future<ApiResponse<List<BankCaseModel>>> getMyComplaints() async {
    final response = await apiClient.get('/banks/cases/my-cases');
    if (response.isSuccess && response.data != null) {
      final List<dynamic> data = response.data is List ? response.data : response.data['data'] ?? [];
      final cases = data.map((json) => BankCaseModel.fromJson(json)).toList();
      return ApiResponse.success(cases);
    }
    return ApiResponse.error(response.error ?? 'Failed to load complaints');
  }

  // Update Status (Staff)
  Future<ApiResponse<bool>> updateCaseStatus(String caseId, String status, {String? notes}) async {
    final response = await apiClient.patch('/banks/cases/$caseId/status', {
      'status': status,
      if (notes != null) 'notes': notes
    });
    if (response.isSuccess) {
      return ApiResponse.success(true);
    }
    return ApiResponse.error(response.error ?? 'Update failed');
  }

  // Get Branch Cases (Staff)
  Future<ApiResponse<List<BankCaseModel>>> getCasesByBranch(String branchId) async {
    final response = await apiClient.get('/banks/branches/$branchId/cases');
    if (response.isSuccess && response.data != null) {
      final List<dynamic> data = response.data is List ? response.data : response.data['data'] ?? [];
      final cases = data.map((json) => BankCaseModel.fromJson(json)).toList();
      return ApiResponse.success(cases);
    }
    return ApiResponse.error(response.error ?? 'Failed to load branch cases');
  }
}
